### PR TITLE
MJX Training Pipeline

### PR DESCRIPTION
**Pull request by michael-lutz**
This PR introduces a new way to massively scale up locomotion training. Building upon Brax, it ultimately uses the MJX physics engine for simulation.

### Structure
Specifically, this PR includes the following directories:

- Envs
- Experiments
- Utils
- (example) Weights
- train.py
- play.py

`Envs` includes two types of Brax environments: DefaultHumanoidEnv and StompyEnv. Each environment includes a main class which implements the Brax environment interface and utilizes MJX for all physics calculations. One important thing to note is that reward functions are **modular**, allowing for quick experimentation.

`Experiments` includes two .yaml files that include sample configurations for model training.

`Utils` include default values, rendering rollouts, etc.

`Weights` currently include default humanoid weights (for locomotion) that should work out of the box

train.py and play.py both integrate with wandb. train.py utilizes the Brax implementation of PPO for now, but can be easily customized if needed.

### Performance Samples
**Training Curves**
Screen Shot 2024-05-22 at 9 12 45 PM
Screen Shot 2024-05-22 at 9 14 44 PM

**Example humanoid robot walking in MJX**
https://github.com/kscalelabs/sim/assets/43460304/8e12b0e6-48ea-4af0-8283-1dc4880767b4

**Humanoid trained in MJX, eval in CPU-based MuJoCo**
https://github.com/kscalelabs/sim/assets/43460304/7f158aeb-6bc9-4056-bd1d-12882adbd13c

### CODE REVIEW

**Review by budzianowski — COMMENTED** [review]
nit: move weights to mjx_gym/tests

File: sim/mjx_gym/envs/__init__.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+from brax import envs
+
+from .default_humanoid_env.default_humanoid import DefaultHumanoidEnv
```
Comment (budzianowski) [inline sim/mjx_gym/envs/__init__.py:3]: avoid relative imports

File: sim/mjx_gym/envs/default_humanoid_env/default_humanoid.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
+import jax
```
Comment (budzianowski) [inline sim/mjx_gym/envs/default_humanoid_env/default_humanoid.py:1]: isort ruff et al., see Makefile for formatting setup

File: sim/mjx_gym/envs/default_humanoid_env/default_humanoid.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+import jax.numpy as jp
+from brax.envs.base import PipelineEnv, State
+from brax.mjx.base import State as mjxState
+from brax.io import mjcf
+import mujoco
+from mujoco import mjx
+from etils import epath
+from .rewards import get_reward_fn
+from utils.default import DEFAULT_REWARD_PARAMS
```
Comment (budzianowski) [inline sim/mjx_gym/envs/default_humanoid_env/default_humanoid.py:10]: nit: move default reward params to rewards.py

File: sim/mjx_gym/envs/stompy_env/stompy.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+import jax
+import jax.numpy as jp
+from brax.envs.base import PipelineEnv, State
+from brax.mjx.base import State as mjxState
+from brax.io import mjcf
+import mujoco
+from mujoco import mjx
+from etils import epath
+import os
+from .rewards import get_reward_fn
```
Comment (budzianowski) [inline sim/mjx_gym/envs/stompy_env/stompy.py:11]: avoid relative imports - sim.mjx_gym.envs.rewards

File: sim/mjx_gym/envs/stompy_env/stompy.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+		self._reward_params = reward_params
+		self._terminate_when_unhealthy = terminate_when_unhealthy
+		self._reset_noise_scale = reset_noise_scale
+		self._exclude_current_positions_from_observation = (exclude_current_positions_from_observation)
+		self._log_reward_breakdown = log_reward_breakdown
+
+		self.reward_fn = get_reward_fn(self._reward_params, self.dt, include_reward_breakdown=True)
+
+	def reset(self, rng: jp.ndarray) -> State:
+		"""Resets the environment to an initial state.
```
Comment (budzianowski) [inline sim/mjx_gym/envs/stompy_env/stompy.py:55]: the formatting looks off.

File: sim/mjx_gym/train.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+import argparse
+from envs import get_env
+from brax.training.agents.ppo import train as ppo
+import functools
+import matplotlib.pyplot as plt
+from datetime import datetime
+from brax import envs
+from brax.io import model
+
+# Parse command-line arguments
```
Comment (budzianowski) [inline sim/mjx_gym/train.py:12]: put this under main function

File: sim/mjx_gym/play.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+from envs import get_env
+from brax.io import model
+from brax.training.agents.ppo import networks as ppo_networks
+from brax.training.acme import running_statistics
+import mediapy as media
+import numpy as np
+from utils.default import DEFAULT_REWARD_PARAMS
+from utils.rollouts import render_mjx_rollout, render_mujoco_rollout
+
+# Parse command-line arguments
```
Comment (budzianowski) [inline sim/mjx_gym/play.py:13]: put this under main function

File: sim/mjx_gym/play.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+print(f'Loaded environment {config.get("env_name", "")} with env.observation_size: {env.observation_size} and env.action_size: {env.action_size}')
+
+# Loading params
+if args.params_path is not None:
+    model_path = args.params_path
+else:
+    model_path = "weights/" + config.get('project_name', 'model') + ".pkl"
+params = model.load_params(model_path)
+normalize = lambda x, y: x
+if config.get('normalize_observations', False):
```
Comment (budzianowski) [inline sim/mjx_gym/play.py:45]: Add comment what's going here

File: sim/mjx_gym/play.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+    normalize = running_statistics.normalize
+policy_network = ppo_networks.make_ppo_networks(env.observation_size, env.action_size, preprocess_observations_fn=normalize)
+inference_fn = ppo_networks.make_inference_fn(policy_network)(params)
+print(f'Loaded params from {model_path}')
+
+# rolling out a trajectory
+render_every = 2
+n_steps = 1000
```
Comment (budzianowski) [inline sim/mjx_gym/play.py:53]: make this a parameter

File: sim/mjx_gym/play.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+if args.use_mujoco:
```
Comment (budzianowski) [inline sim/mjx_gym/play.py:54]: Is there actually any difference?

File: sim/mjx_gym/play.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    images = render_mujoco_rollout(env, inference_fn, n_steps, render_every)
+else:
+    images = render_mjx_rollout(env, inference_fn, n_steps, render_every)
+print(f'Rolled out {len(images)} steps')
+
+# render the trajectory
```
Comment (budzianowski) [inline sim/mjx_gym/play.py:60]: Make it optional

File: sim/mjx_gym/train.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+    batch_size=config['batch_size'],
+    seed=config['seed']
+)
+
+x_data = []
+y_data = []
+ydataerr = []
+times = [datetime.now()]
+
+max_y, min_y = 13000, 0
```
Comment (budzianowski) [inline sim/mjx_gym/train.py:73]: I don't think this is used anywhere?

**Review by budzianowski — CHANGES_REQUESTED** [review]
Was stompy env tested at all? If not, let's add it in the next PR since this one is already getting big.

File: sim/mjx_gym/play.py (hunk 10 above)
Comment (michael-lutz) [reply sim/mjx_gym/play.py:62]: Good question!

On performance: MJX takes a while to initialize, very quick to run on GPU. MuJoCo has a quicker "cold start" equivalent, but actual rollouts are slightly slower.

On model performance: due to slightly different physical dynamics, models are slightly less performant on MuJoCo if trained on MJX. See the second vs first video included in the description of the PR.

File: sim/mjx_gym/play.py (hunk 11 above)
Comment (michael-lutz) [reply sim/mjx_gym/play.py:68]: I was envisioning this script to be dedicated primarily to rendering, as we can access all other metrics directly in training logs I believe. Do you think it makes sense to make optional?

File: sim/mjx_gym/envs/default_humanoid_env/default_humanoid.py [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+        self._exclude_current_positions_from_observation = exclude_current_positions_from_observation
+        self._log_reward_breakdown = log_reward_breakdown
+
+        self.reward_fn = get_reward_fn(self._reward_params, self.dt, include_reward_breakdown=True)
+
+    def reset(self, rng: jp.ndarray) -> State:
+        """Resets the environment to an initial state.
+
+        Args:
+                rng: Random number generator seed.
```
Comment (budzianowski) [inline sim/mjx_gym/envs/default_humanoid_env/default_humanoid.py:58]: nit - formatting still off here
```suggestion
     Args:
            rng: Random number generator seed.
```

File: sim/mjx_gym/envs/default_humanoid_env/rewards.py [hunk 14]
```diff
@@ -0,0 +1,111 @@
+from typing import Callable, Dict, Tuple
```
Comment (budzianowski) [inline sim/mjx_gym/envs/default_humanoid_env/rewards.py:1]: Good practice is to add even one line comment what the file is about.

File: sim/mjx_gym/envs/default_humanoid_env/rewards.py [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+from brax.mjx.base import State as mjxState
+
+DEFAULT_REWARD_PARAMS = {
+    "rew_forward": {"weight": 1.25},
+    "rew_healthy": {"weight": 5.0, "healthy_z_lower": 1.0, "healthy_z_upper": 2.0},
+    "rew_ctrl_cost": {"weight": 0.1},
+}
+
+
+def get_reward_fn(
```
Comment (budzianowski) [inline sim/mjx_gym/envs/default_humanoid_env/rewards.py:14]: This could be used across all envs?

File: sim/mjx_gym/envs/default_humanoid_env/rewards.py [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+    Args:
+        state: Current state.
+        action: Action taken.
+        next_state: Next state.
+        dt: Time step.
+        params: Reward parameters.
+    Returns:
+        A float wrapped in a jax array.
+    """
+    xpos = state.subtree_com[1][0]  # TODO: include stricter typing than mjxState to avoid this type error
```
Comment (budzianowski) [inline sim/mjx_gym/envs/default_humanoid_env/rewards.py:56]: Make it more explicit what you are loading

File: sim/mjx_gym/envs/default_humanoid_env/rewards.py [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+    next_xpos = next_state.subtree_com[1][0]
+    velocity = (next_xpos - xpos) / dt
+    forward_reward = params["weight"] * velocity
+
+    return forward_reward, jp.array(1.0)  # TODO: ensure everything is initialized in a size 2 array instead...
```
Comment (budzianowski) [inline sim/mjx_gym/envs/default_humanoid_env/rewards.py:61]: What's the logic behind 1.0?

File: sim/mjx_gym/envs/stompy_env/rewards.py [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+    velocity = (next_xpos - xpos) / dt
+    forward_reward = params["weight"] * velocity
+
+    return forward_reward, jp.array(1.0)  # TODO: ensure everything is initialized in a size 2 array instead...
+
+
+def healthy_reward_fn(
+    state: mjxState, action: jp.ndarray, next_state: mjxState, dt: jax.Array, params: Dict[str, float]
+) -> Tuple[jp.ndarray, jp.ndarray]:
+    """Reward function for staying healthy.
```
Comment (budzianowski) [inline sim/mjx_gym/envs/stompy_env/rewards.py:67]: rename upright _reward

File: sim/mjx_gym/envs/stompy_env/stompy.py [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+        self._log_reward_breakdown = log_reward_breakdown
+
+        self.reward_fn = get_reward_fn(self._reward_params, self.dt, include_reward_breakdown=True)
+
+    def reset(self, rng: jp.ndarray) -> State:
+        """
+        Resets the environment to an initial state.
+
+        Args:
+                        rng: Random number generator seed.
```
Comment (budzianowski) [inline sim/mjx_gym/envs/stompy_env/stompy.py:55]: nit - weird tabs

File: sim/mjx_gym/play.py [hunk 20]
```diff
@@ -0,0 +1,81 @@
+import argparse
```
Comment (budzianowski) [inline sim/mjx_gym/play.py:1]: Add example how to run it

File: sim/mjx_gym/train.py [hunk 21]
```diff
@@ -0,0 +1,80 @@
+import argparse
```
Comment (budzianowski) [inline sim/mjx_gym/train.py:1]: Add example how to run it

**Review by budzianowski — APPROVED** [review]
small nits but looks good for a start!

File: sim/mjx_gym/envs/default_humanoid_env/rewards.py (hunk 17 above)
Comment (michael-lutz) [reply sim/mjx_gym/envs/default_humanoid_env/rewards.py:61]: jp.array because we want to keep everything in jax. The 1.0 itself operates like a boolean operator (doesn't change the "healthiness" until a 0 comes around)
